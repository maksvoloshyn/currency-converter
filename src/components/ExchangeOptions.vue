<template>
    <div class="flex flex-col sm:items-center">
        <div v-for="(option, index) in exchangeOptions" :key="index" class="flex items-center py-3 w-11/12 sm:w-auto">
            <exchange-option
                :currencies="currencies"
                v-bind="option"
                @update:amount="updateBaseAmount($event, option.code)"
                @update:code="updateCurrencyForExchange($event, index)"
            ></exchange-option>
            <button
                v-show="canRemoveExchangeOption"
                class="ml-3 p-2 rounded-full"
                aria-label="Remove currency"
                @click="removeExchangeOption(index)"
            >
                &#10006;
            </button>
        </div>

        <button class="bg-blue-600 border font-bold px-3 py-2 rounded-lg text-white mt-4" @click="addExchangeOption">
            Add currency
        </button>
    </div>
</template>

<script>
    import ExchangeOption from '@/components/ExchangeOption';

    const DEFAULT_EXCHANGE_CURRENCY = 'USD';

    export default {
        name: 'exchange-options',

        components: {
            'exchange-option': ExchangeOption,
        },

        props: {
            rates: {
                type: Object,
                default: () => ({}),
            },

            baseCurrency: {
                type: String,
            },

            currencies: {
                type: Array,
                default: () => [],
            },
        },

        data() {
            const defaultExchangeCurrency = this.rates[DEFAULT_EXCHANGE_CURRENCY]
                ? DEFAULT_EXCHANGE_CURRENCY
                : this.currencies[0].code;

            return {
                pairs: [],
                currenciesForExchange: [this.baseCurrency, defaultExchangeCurrency],
                amountInBaseCurrency: 1,
            };
        },

        computed: {
            exchangeOptions() {
                return this.currenciesForExchange.map(currency => ({
                    code: currency,
                    amount: (currency && this.amountInBaseCurrency * this.rates[currency]) || 0,
                }));
            },

            canRemoveExchangeOption() {
                return this.exchangeOptions.length > 2;
            },
        },

        methods: {
            updateBaseAmount(value, code) {
                const normalisedValue = typeof value === 'string' ? parseFloat(value) : value;

                this.amountInBaseCurrency = normalisedValue / this.rates[code];
            },

            updateCurrencyForExchange(code, index) {
                this.currenciesForExchange.splice(index, 1, code);
            },

            addExchangeOption() {
                this.currenciesForExchange.push(null);
            },

            removeExchangeOption(index) {
                this.currenciesForExchange.splice(index, 1);
            },
        },
    };
</script>

<style scoped></style>
