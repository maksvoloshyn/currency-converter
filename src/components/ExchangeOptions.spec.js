import {shallowMount} from '@vue/test-utils';
import ExchangeOptions from '@/components/ExchangeOptions';

const ExchangeOptionStub = {
    template: '<span>{{code}} - {{amount}}</span>',
    props: ['currencies', 'code', 'amount'],
};

function factory() {
    return shallowMount(ExchangeOptions, {
        propsData: {
            currencies: currenciesMock,
            rates: ratesMock,
        },
        stubs: {
            'exchange-option': ExchangeOptionStub,
        },
    });
}

const currenciesMock = [{UAH: 'Ukrainian hryvna', EUR: 'Euro', USD: 'United States Dollar'}];
const ratesMock = {
    EUR: 1,
    UAH: 33,
    USD: 1.2,
};

describe('ExchangeOptions', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = factory();
    });

    it('renders correct template', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    describe('when `Add currency button is clicked`', () => {
        beforeEach(() => {
            return wrapper.find('#add-currency').trigger('click');
        });

        it('renders a new ExchangeOption component with nullish values', () => {
            const newOption = wrapper.findAllComponents(ExchangeOptionStub).at(2);

            expect(newOption.exists()).toBe(true);
            expect(newOption.props()).toEqual({
                currencies: currenciesMock,
                code: null,
                amount: 0,
            });
        });

        it('shows remove buttons for each option', () => {
            const removeButtons = wrapper.findAll('[aria-label="Remove currency"]');

            expect(removeButtons.length).toBe(3);

            removeButtons.wrappers.forEach(wrapper => {
                expect(wrapper.isVisible()).toBe(true);
            });
        });

        describe('when ExchangeOption emits `update:code`', () => {
            beforeEach(() => {
                const target = wrapper.findAllComponents(ExchangeOptionStub).at(2);

                target.vm.$emit('update:code', 'UAH');
            });

            it('updates its code and amount props to meet new currency', async () => {
                expect(wrapper.findAllComponents(ExchangeOptionStub).at(2).props()).toEqual({
                    currencies: currenciesMock,
                    code: 'UAH',
                    amount: 33,
                });
            });

            describe('when ExchangeOption emits `update:amount`', () => {
                it('updates amounts for all options to meet new exchange value', async () => {
                    const [optionA, optionB, optionC] = wrapper.findAllComponents(ExchangeOptionStub).wrappers;

                    await optionA.vm.$emit('update:amount', 10);

                    expect(optionA.props()).toEqual({
                        currencies: currenciesMock,
                        code: 'EUR',
                        amount: 10,
                    });

                    expect(optionB.props()).toEqual({
                        currencies: currenciesMock,
                        code: 'USD',
                        amount: 12,
                    });

                    expect(optionC.props()).toEqual({
                        currencies: currenciesMock,
                        code: 'UAH',
                        amount: 330,
                    });
                });
            });

            describe('when remove button is clicked', () => {
                beforeEach(() => {
                    wrapper.findAll('[aria-label="Remove currency"]').at(1).trigger('click');
                });

                it('removes relevant option', () => {
                    const options = wrapper.findAllComponents(ExchangeOptionStub);

                    expect(options.length).toBe(2);
                    expect(options.at(0).props('code')).toBe('EUR');
                    expect(options.at(1).props('code')).toBe('UAH');
                });

                it('hides remove buttons', () => {
                    const removeButtons = wrapper.findAll('[aria-label="Remove currency"]');

                    expect(removeButtons.length).toBe(2);

                    removeButtons.wrappers.forEach(wrapper => {
                        expect(wrapper.isVisible()).toBe(false);
                    });
                });
            });
        });
    });
});
