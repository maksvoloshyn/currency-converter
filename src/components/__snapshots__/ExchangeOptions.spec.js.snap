// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeOptions renders correct template 1`] = `
<div
  class="flex flex-col sm:items-center"
>
  <div
    class="flex items-center py-3 w-11/12 sm:w-auto"
  >
    <span>
      EUR - 1
    </span>
     
    <button
      aria-label="Remove currency"
      class="ml-3 p-2 rounded-full"
      style="display: none;"
    >
      
            ✖
        
    </button>
  </div>
  <div
    class="flex items-center py-3 w-11/12 sm:w-auto"
  >
    <span>
      USD - 1.2
    </span>
     
    <button
      aria-label="Remove currency"
      class="ml-3 p-2 rounded-full"
      style="display: none;"
    >
      
            ✖
        
    </button>
  </div>
   
  <button
    class="bg-blue-600 border font-bold px-3 py-2 rounded-lg text-white mt-4"
    id="add-currency"
  >
    
        Add currency
    
  </button>
</div>
`;
